File: XC9572XL.pl
#!/usr/bin/perl

#rename this file from .txt to .pl

#TODO
#Support all commands in SVF specification
#Implement thorough error handling
#Verify TDO values. Currently hardware supports this but not implemented in this script

#Numato Lab www.numato.com, www.numato.cc
#License : http://creativecommons.org/licenses/by-sa/2.0/

# Determin operating system
$Os = $^O;

use Time::HiRes(usleep);
use POSIX qw(ceil floor);

if($Os == "MSWin32"){
	use Win32::SerialPort;
}else{
	#Use Device::SerialPort for linux
	print "Only Win32 Operating Systems supported";
}

my $LineNumber = 0;
my $TdiLength = 0;

$portName = $ARGV[0];
$FileName = $ARGV[1];

###########################################################################
#                         Chipmunk APIs                                   #
###########################################################################

%JTAG_STATES = (
	'TL_RESET' => 0x0,
	'IDLE' => 0x1,
	'DRSELECT' => 0x9,
	'DRCAPTURE' => 0xa,
	'DRSHIFT' => 0xb,
	'DREXIT1' => 0xc,
	'DRPAUSE' => 0xd,
	'DREXIT2' => 0xe,
	'DRUPDATE' => 0xf,
	'IRSELECT' => 0x2,
	'IRCAPTURE' => 0x3,
	'IRSHIFT' => 0x4,
	'IREXIT1' => 0x5,
	'IRPAUSE' => 0x6,
	'IREXIT2' => 0x7,
	'IRUPDATE' => 0x8
);

sub OpenPort($)
{
	my $strPortName = shift;
	
	# Open serial port and configure
	$port = new Win32::SerialPort($strPortName, quiet)
			   || die "Could not open the port specified";
			   
	$port->baudrate(9600);
	$port->parity("none");
	$port->databits(8);
	$port->stopbits(1);
	$port->handshake("none");
	$port->buffers(4096, 4096); 
	$port->read_interval(100);   
	$port->read_char_time(5);  
	$port->read_const_time(100); 
	$port->write_char_time(5);
	$port->write_const_time(100);
	$port->lookclear();
	
	return $port;
}

sub SendAndReceive($$$$$)
{
	my $port = shift;
	my $SendData = shift;
	my $SendLen = shift;
	my $ReceiveData = shift;
	my $ReceiveLen = shift;
	
	$port->lookclear();
	
	$BytesWritten = $port->write($SendData);
	if($BytesWritten !=  $SendLen)
	{
		return 0;
	}

	($count, $$ReceiveData) = $port->read($ReceiveLen);
	if($count != $ReceiveLen)
	{
		return 0;		
	}
	
	if($$ReceiveData =~ /\?/)
	{
		print "Chipmunk returned error at line number $LineNumber\n";
		$port->close();
		die;
	}

	return $count ;
}

sub EchoOn()
{
	SendAndReceive($port, "echo on\r", 9, NULL, NULL);
}

sub EchoOff()
{
	SendAndReceive($port, "echo off\r", 10, NULL, NULL);
	#Dummy read to clear input buffer. LookClear seems to be not working
	$port->read(0xff);
}

#Get firmware version
sub GetVersion($)
{
	my $port = shift;
	my $tmpString;
	
	if(SendAndReceive($port, "ver\r",4, \$tmpString, 8))
	{
		return $tmpString;
	}
	
	return NULL;
}

#Set enddr
sub Enddr($$)
{
	my $port = shift;
	my $state = shift;
	my $tmpString;
	
	SendAndReceive($port, "enddr ".sprintf("%x", $state)."\r",8, \$tmpString, 8);
}

#Set enddir
sub Endir($$)
{
	my $port = shift;
	my $state = shift;
	my $tmpString;
	
	SendAndReceive($port, "endir ".sprintf("%x", $state)."\r",8, \$tmpString, 8);
}

#Put the JTAG TAP to reset state
sub Reset($)
{
	my $port = shift;
	my $tmpString;
	
	SendAndReceive($port, "reset\r",6, \$tmpString, 0);
}

#Advance TAP state machine
sub TapAdvance($$)
{
	my $port = shift;
	my $tms = shift;
	my $tmpString;
	
	SendAndReceive($port, "a ".sprintf("%x", $tms)."\r",4, \$tmpString, 2);
}

#Go to a specific TAP state
sub GoToState($$)
{
	my $port = shift;
	my $state = shift;
	my $tmpString;
	
	SendAndReceive($port, "g ".sprintf("%x", $state)."\r",4, \$tmpString, 2);
}

sub GetState($)
{
	my $port = shift;
	my $tmpString;
	
	if(SendAndReceive($port, "q\r",2 , \$tmpString, 1) != 0)
	{
		return $tmpString;
	}
	return NULL;
}

sub SelectIR($)
{
	my $port = shift;
	my $tmpString;
	my $state;
	
	SendAndReceive($port, "i\r",2 , \$tmpString, 1);
	
	$state = GetState($port);
	if($state ne "4")
	{
		GoToState($port, 0);
		SendAndReceive($port, "i\r",2 , \$tmpString, 1);
	}
	
	$state = GetState($port);
	return $state;
}

sub SelectDR($)
{
	my $port = shift;
	my $tmpString;
	my $state;
	
	SendAndReceive($port, "d\r",2 , \$tmpString, 1);
	
	$state = GetState($port);
	if($state ne "B")
	{
		GoToState($port, 0);
		SendAndReceive($port, "d\r",2 , \$tmpString, 1);
	}
	
	$state = GetState($port);
	return $state;
}

#Scan the selected register
sub Scan($$$$)
{
	my $port = shift;
	my $Tdi = shift;
	my $TdiLen = shift;
	my $ExitState = shift;
	
	if($TdiLen > 32)
	{
		return NULL;
	}
	
	if($ExitState == 0)
	{
		if(SendAndReceive($port, "s ". substr($Tdi, 0, ceil($TdiLen/4)) ." ". sprintf("%02x", $TdiLen)."\r", ceil($TdiLen/4)+6, \$tmpString, (ceil($TdiLen/4))))
		{
			return $tmpString;
		}
		else
		{
			return NULL;
		}
	}
	else
	{
		if(SendAndReceive($port, "x ". substr($Tdi, 0, ceil($TdiLen/4)) ." ". sprintf("%02x", $TdiLen)."\r", ceil($TdiLen/4)+6, \$tmpString, (ceil($TdiLen/4))))
		{
			return $tmpString;
		}
		else
		{
			return NULL;
		}
	}
}

#Runtest
sub Runtest($$)
{
	my $port = shift;
	my $cycles = shift;
	my $tmpString;
	
	SendAndReceive($port, "r ".sprintf("%x", $cycles)."\r",4 , \$tmpString, 0);
	
	#Wait a ltille bit before issuing next command
	usleep($cycles/2);
}

#Load SVF file and process
sub RunSVF($$)
{
	my $port = shift;
	my $FileName = shift;
	
	print "Processing SVF, please wait\n";
	
	open(SVF, "<$FileName");
	
	while(!eof(SVF))
	{
		my $line = <SVF>;
		$LineNumber++;
		
		print "Processing line $LineNumber\r";
		
		#Check and strip any comments
		if($line =~ /\/\//i)
		{
			$line = substr($line, 0, @- - 1);
		}
		
		#strip extra white spaces
		$line =~ s/[\h\v]+/ /g;
		
		if($line =~ /TRST/i)
		{
			#Ignore
		}
		elsif($line =~ /ENDIR/i)
		{
			if($line !~ /;/)
			{
				print "Error at line no. $LineNumber";
				return;
			}
			
			EndirHandler($port, $line);
		}
		elsif($line =~ /ENDDR/i)
		{
			if($line !~ /;/)
			{
				print "Error at line no. $LineNumber";
				return;
			}
			
			EnddrHandler($port, $line);
		}
		elsif($line =~ /STATE/i)
		{
			if($line !~ /;/)
			{
				print "Error at line no. $LineNumber";
				return;
			}
			
			StateHandler($port, $line);
		}
		elsif($line =~ /RUNTEST/i)
		{
			if($line !~ /;/)
			{
				print "Error at line no. $LineNumber, missing semicolon";
				return;
			}
			
			RuntestHandler($port, $line);
		}
		elsif($line =~ /SIR/i)
		{
			SirHandler($port, $line);
		}
		elsif($line =~ /SDR/i)
		{
			#Sometimes SDR TDI data can be very long spanning multiple lines
			my $end = 0;
			while(!$end)
			{
				if($line =~ /\)/)
				{
					SdrHandler($port, $line, 1);
					$end = 1;
				}
				else
				{
					SdrHandler($port, $line, 0);
					$line = <SVF>;
					$LineNumber++;
				}
			}	
		}
	}
	
	close SVF;
	print "\nDone\n";
	return;
}

sub SirHandler($$)
{
	my $port = shift;
	my $line = shift;

	SelectIR($port);
	
	$line =~ /SIR *([0-9]*) *TDI *\(([0-9a-fA-F]+)\)/i;
	Scan($port, $2, $1, 1);
}

sub SdrHandler($$$)
{
	my $port = shift;
	my $line = shift;
	my $end = shift;
	
	my $LineLength = 0;
	my $i = 0;
	
	if($line =~ /SDR +([0-9]+)/i)
	{
		$TdiLength = $1;
		
		$line =~ /TDI +\((.+)/i;		
		$line = $1;
	}

	if($end)
	{
		$line =~ /([0-9a-fA-F]*)\)/i;
		$line = $1;
	}

	$line =~ s/\s+$//;
	
	SelectDR($port);
	
	$LineLength = length($line) * 4;
	
	if($LineLength > $TdiLength)
	{
		$line = substr($line, floor(($LineLength - $TdiLength)/4), $LineLength/4);		
		$LineLength = $TdiLength;
	}
	
	my $AdjustedLineLength = ceil($LineLength/4);
	
	while($LineLength)
	{
		if($LineLength >= 32)
		{
			$i = $i + 8;
			Scan($port, substr($line, -$i, 8) , 32, 0);
			$LineLength = $LineLength - 32;
			$TdiLength = $TdiLength - 32;
		}
		else
		{
			Scan($port, substr($line,-($AdjustedLineLength),$AdjustedLineLength - $i) , $LineLength, $end);
			$LineLength = 0;
		}
	}

}

sub RuntestHandler($$)
{
	my $port = shift;
	my $line = shift;
	$line =~ /\s+(.*)\s*;/i;
	
	if($line !~ /TCK/i)
	{
		print "Error at line no. $LineNumber, only TCK is supported with RUNTEST";
	}
	
	$line =~ /\s+(.*)\s*;/i;
	
	Runtest($port , $1);
}

sub EndirHandler($$)
{
	my $port = shift;
	my $line = shift;
	$line =~ /\s+(.*)\s*;/i;
	Endir($port , $JTAG_STATES{$1});
}

sub EnddrHandler($$)
{
	my $port = shift;
	my $line = shift;
	$line =~ /\s+(.*)\s*;/i;
	Enddr($port , $JTAG_STATES{$1});
}

sub StateHandler($$)
{
	my $port = shift;
	my $line = shift;
	$line =~ /\s+(.*)\s*;/i;
	GoToState($port , $JTAG_STATES{$1});
}


##################################################################################

$serport = OpenPort($portName);

EchoOff();
RunSVF($serport, $FileName);

$serport->close();

File: XC9572XL.py
'''
Created on Jul 12, 2013
Last Modified on Aug 17, 2013

@author: Rohit
'''

import sys                             # for command line arguments 
import serial                         # for serial port access
import re                               # regex
from math import *              # for floor() and ceil()
import time                           # for usleep() replacement

''' 
determine OS 
$Os = $^O;                        # ^O is contains operating system value

use Time::HiRes(usleep);
use POSIX qw(ceil floor);

if($Os == "MSWin32"){
        use Win32::SerialPort;
}else{
        #Use Device::SerialPort for linux
        print "Only Win32 Operating Systems supported";
}
'''

usleep = lambda x: time.sleep( x / 1000000.0 ) #usleep() derivation from time.sleep()

LineNumber = 0
TdiLength = 0

portName = sys.argv[1]          # "COM45"
FileName = sys.argv[2]                 #"s.svf"

#JTAG_STATES==> Changed TL_RESET state to RESET
JTAG_STATES = {           'RESET' :  0x0,
                                       'IDLE' : 0x1,
                                       'DRSELECT' : 0x9,
                                       'DRCAPTURE' : 0xa,
                                       'DRSHIFT' : 0xb,
                                       'DREXIT1' : 0xc,
                                       'DRPAUSE' : 0xd,
                                       'DREXIT2' : 0xe,
                                       'DRUPDATE' : 0xf,
                                       'IRSELECT' : 0x2,
                                       'IRCAPTURE' : 0x3,
                                       'IRSHIFT' : 0x4,
                                       'IREXIT1' : 0x5,
                                       'IRPAUSE' : 0x6,
                                       'IREXIT2' : 0x7,
                                       'IRUPDATE' : 0x8} 

#print JTAG_STATES

def OpenPort( strPortName ):
        ''' '''
        port = serial.Serial()
        port.baudrate = 9600
        port.parity = serial.PARITY_NONE
        port.bytesize = serial.EIGHTBITS
        port.stopbits = serial.STOPBITS_ONE 
        port.timeout = 5
        port.port = strPortName
        port.open()
        return port

        '''
        $port -> parity( "none" )
        $port -> databits( 8 )
        $port -> stopbits( 1 )
        $port -> handshake( "none" )
        $port -> buffers( 4096, 4096 )
        $port -> read_interval( 100 )  
        $port -> read_char_time( 5 )
        $port -> read_const_time( 100 )
        $port -> write_char_time( 5 )
        $port -> write_const_time( 100 )
        $port -> lookclear()'''
        
        return port

def SendAndReceive( port = None , SendData = None, SendLen = None, ReceiveData = None, ReceiveLen = None ):
        ''' '''
        if port is None:
                print "NULL port provided"
                return
        
        port.flush()            # $port->lookclear();
        print "To Write: " + SendData
        BytesWritten = port.write( SendData )
        if BytesWritten != SendLen:
                return 0
       
        ReceiveData = port.read( ReceiveLen )
        print "Read Data: " + ReceiveData
		
        count = len( ReceiveData )
        if count != ReceiveLen:
                return 0
        
		
        if re.search( "\?", ReceiveData ):
                print "Chipmunk returned error at line number $LineNumber\n"
                port.close()
                exit()
        return count, ReceiveData
        
def EchoOn( port = None ):
        ''' '''
        SendAndReceive( port, "echo on\r", 9, None, None )
        
def EchoOff( port = None ):
        ''' '''
        SendAndReceive( port, "echo off\r", 10, None, None )
        port.flush()
        
def GetVersion( _port = None ):
        ''' '''
        port = _port
        tmpString = ""
        
        len, tmpString = SendAndReceive( port, "ver\r", 4, tmpString, 8 )
        if  len:
                return tmpString
        return None

def Enddr( _port, _state ):
        ''' '''
        port = _port
        state = _state
        tmpString = ""
        len = 0
        tmpString = SendAndReceive( port, "enddr " + hex( state )[2:] + "\r", 8, tmpString, 8 )
        print tmpString
        return tmpString
        
def Endir( _port, _state ):
        ''' '''
        port = _port
        state = _state
        tmpString = ""
        len = 0
        tmpString = SendAndReceive( port, "endir " + hex( state )[2:] + "\r", 8, tmpString, 8 )
        print tmpString
        return tmpString
 
def Reset( _port ):
        ''' '''
        port = _port
        tmpString = ""
        len = 0
        tmpString = SendAndReceive( port, "reset\r", 6, tmpString, 0 )
        print tmpString
        return tmpString
        
def TapAdvance( _port, _tms ):
        ''' '''
        port = _port
        tmpString = ""
        tms = _tms
        len = 0
        
        tmpString = SendAndReceive( port, "a " + hex( tms )[2:] + "\r", 4, tmpString, 2 )
        print tmpString
        return tmpString
        
def GoToState( _port, _state ):
        ''' '''
        port = _port
        state = _state
        tmpString = ""
        len = 0
        tmpString = SendAndReceive( port, "g " + hex( state )[2:] + "\r", 4, tmpString, 2 )
        print tmpString
        return tmpString
         
def GetState( _port ):
        ''' '''
        port = _port
        tmpString = ""
        len = 0
        len, tmpString = SendAndReceive( port, "q\r", 2 , tmpString, 1 )
        if  len != 0:
                return tmpString
        return None
       
def SelectIR( _port ):
        ''' '''
        port = _port
        tmpString = ""
        state = None
        
        SendAndReceive( port, "i\r", 2 , tmpString, 1 )
        state = GetState( port )
        if state != "4":
                GoToState( port, 0 )
                SendAndReceive( port, "i\r", 2 , tmpString, 1 )
        
        
        state = GetState( port )
        return state
        
def SelectDR( _port ):
        ''' '''
        port = _port
        tmpString = ""
        state = None
        
        SendAndReceive( port, "d\r", 2 , tmpString, 1 )
        state = GetState( port )
        if state != "B":
                GoToState( port, 0 )
                SendAndReceive( port, "d\r", 2 , tmpString, 1 )
                
        
        state = GetState( port )
        return state

def Scan( _port, _tdi, _tdilen, _exitstate ):
        ''' '''
        port = _port
        Tdi = _tdi
        TdiLen = int(_tdilen)
        ExitState = _exitstate
        tmpString = ""
        len = 0
        if TdiLen > 32:
                return None
        
        if  ExitState == 0 :
                len, tmpString = SendAndReceive( port, "s " + substr( Tdi, 0, int( ceil( TdiLen / 4.0 ) ) ) + " " + ( "%02x" % TdiLen ) + "\r", int( ceil( TdiLen / 4.0 ) ) + 6, tmpString, int( ceil( TdiLen / 4.0 ) ) )
                if len:
                        return tmpString
                else:
                        return None
        else:
                print "hulla", substr( Tdi, 0, int( ceil( TdiLen / 4.0 ) ) )
                len, tmpString = SendAndReceive( port, "x " + substr( Tdi, 0, int( ceil( TdiLen / 4.0 ) ) ) + " " + ( "%02x" % TdiLen ) + "\r", int( ceil( TdiLen / 4.0 ) ) + 6, tmpString, int( ceil( TdiLen / 4.0 ) ) ) 
                if len:
                        return tmpString
                else:
                        None        

def Runtest( _port, _cycles ):
        ''' '''
        port = _port
        cycles = int(re.sub("TCK", "", _cycles))
        tmpString = ""
        
        SendAndReceive( port, "r " + ( "%x" % cycles ) + "\r", 4 , tmpString, 0 )
        #Wait a ltille bit before issuing next command
        usleep( cycles / 2 )
        
def RunSVF( _portName, _FileName ):
        port = _portName
        FileName = _FileName
        portName = port.port
        
        print "Processing SVF, please wait:\nPort: " + portName + "\nFilename: " + FileName + "\n\n****************\n"
        
        SVF = open( FileName, 'r' )
        SVF_enumerate = enumerate( SVF )
        
        for i, line in SVF_enumerate:
                #print str(i+1) +" : " + line,
                #print "\n\n____________________________________\nProcessing line " + str( i + 1 ) + "\r\nLine: " + line
                #time.sleep(0.005)
                #Check and strip any comments
                comment_found = re.search( "\/\/", line )
                if comment_found:
                        #print "COMMENT FOUND!!"
                        line = line[0:comment_found.start()]  #line without comment part
                        #print "Line is : " + line
                
                #strip extra white spaces
                line = re.sub( r"\s+", " ", line )
                #print "Line w/o whitespace: " + line + "hello"
                
                if re.search( "TRST", line, re.IGNORECASE ):
                        #Ignore
                        print "found /TRST/i...ignored"
                
                elif re.search( "ENDIR", line, re.IGNORECASE ):
                        if not re.search( ";", line ):
                                print "Error at line no. " + ( i + 1 )
                                return
                        
                        #print "EndirHandler(" + port + ", " + line + ")"
                        EndirHandler( port , line )
                
                elif re.search( "ENDDR", line, re.IGNORECASE ):
                        if not re.search( ";", line ):
                                print "Error at line no. " + ( i + 1 )
                                return
                        
                        #print "EnddrHandler(" + port + ", " + line + ")"
                        EnddrHandler( port, line )
                                
                elif re.search( "STATE", line, re.IGNORECASE ):
                        if not re.search( ";", line ):
                                print "Error at line no. " + ( i + 1 )
                                return
                        
                        #print "StateHandler(" + port + ", " + line + ")"
                        StateHandler( port, line )
                                
                elif re.search( "RUNTEST", line, re.IGNORECASE ):
                        if not re.search( ";", line ):
                                print "Error at line no. " + ( i + 1 )
                                return
                        
                        #print "RuntestHandler(" + port + ", " + line + ")"
                        RuntestHandler( port, line )

                elif re.search( "SIR", line, re.IGNORECASE ):
                        #if not re.search(";", line):
                        #        print "Error at line no. " + (i+1)
                        #        return
                        
                        #print "SirHandler(" + port + ", " + line + ")"
                        SirHandler( port, line )

                elif re.search( "SDR", line, re.IGNORECASE ):
                        
                        #Sometimes SDR TDI data can be very long spanning multiple lines
                        end = 0
                        while not end:
                                if re.search( "\)", line ):
                                        #print "SdrHandler(" + port + ", " + line + ", 1)"
                                        SdrHandler( port, line, 1 )
                                        end = 1
                                else:
                                        #print "SdrHandler(" + port + ", " + line + ", 0)"
                                        SdrHandler( port , line , 0 )
                                        i, line = SVF_enumerate.next()        #increment line, and linenumber
                                                       
        SVF.close()
        print "\nDone\n"
        return       

def SirHandler( _port, _line ):
        port = _port
        line = _line
        portName = port.port
        #print "\n\n" + portName + " -- " + line + " ||"
         
        
        #print "SelectIR(" + port + ") \n"
        SelectIR( port )
        
        regmatch = re.search( r"SIR *([0-9]*) *TDI *\(([0-9a-fA-F]+)\)", line, re.IGNORECASE )
        
        print "Scan(" + portName + ", " + regmatch.group( 2 ) + ", " + regmatch.group( 1 ) + ", 1);" 
        Scan( port, regmatch.group( 2 ), regmatch.group( 1 ), 1 )
             
def SdrHandler( _port, _line, _end ):
        port = _port
        line = _line
        end = _end
        portName = port.port
        print line
        LineLength = 0
        i = 0
        
        regmatch = re.search( r"SDR +([0-9]+)", line, re.IGNORECASE )
        if regmatch:
                TdiLength = int( regmatch.group( 1 ) )
                regmatch = re.search( r"TDI +\((.+)", line, re.IGNORECASE )
                line = regmatch.group( 1 )
        
        if end:
                regmatch = re.search( r"([0-9a-fA-F]*)\)", line, re.IGNORECASE )
                line = regmatch.group( 1 )
        
        line = re.sub( r"\s+$", "", line )
        #print "SelectDR(" + port + ");" 
        SelectDR( port )
        
        LineLength = len( line ) * 4
        #print " Print1: ",line, LineLength, TdiLength
        if LineLength > TdiLength:
                #print "Truce"
                line = substr( line, int( floor( ( LineLength - TdiLength ) / 4.0 ) ), int( LineLength / 4 ) )
                LineLength = TdiLength
        
        AdjustedLineLength = int( ceil( LineLength / 4.0 ) )
        #print " Print2: ",line, LineLength, TdiLength, AdjustedLineLength
        
        while LineLength:
                if LineLength >= 32:
                        #print "hulla"
                        i = i + 8
                        #print "Scan(" + port + ", substr(" + line + ", " + str( -i ) + ", 8) , 32, 0);"
                        print "Scan(" + portName + ", " + substr( line, -i, 8 ) + " , 32, 0);"
                        Scan( port, substr( line, -i, 8 ) , 32, 0 )
                        LineLength = LineLength - 32
                        TdiLength = TdiLength - 32
                else:
                        #print "hallooo"
                        #print "Scan(" + port + ", substr(" + line + ",-(" + \
                        #                str( AdjustedLineLength ) + ")," + str( AdjustedLineLength ) + \
                        #                " - " + str( i ) + " , " + str( LineLength ) + ", " + str( end ) + ");"
                                        
                        print "Scan(" + portName + ", " + substr( line, -AdjustedLineLength, AdjustedLineLength - i ) + \
                                        " , " + str( LineLength ) + ", " + str( end ) + ");"
                        Scan( port, substr( line, -AdjustedLineLength, AdjustedLineLength - i ) , LineLength, end )
                        LineLength = 0

def RuntestHandler( _port, _line ):
        port = _port
        line = _line
        #portName = port.port
        
        print line
        
        if not re.search( "TCK", line, re.IGNORECASE ) :
                print "Error at line no. LineNumber, only TCK is supported with RUNTEST"
        
        regmatch = re.search( r"\s+(.*)\s*;", line, re.IGNORECASE )
        #print "Runtest(" + port + ", " + regmatch.group( 1 ) + ");" 
        Runtest( port, regmatch.group( 1 ) )

def EndirHandler( _port, _line ):
        port = _port
        line = _line
        #portName = port.port

        print line
        
        regmatch = re.search( r"\s+(.*)\s*;", line, re.IGNORECASE )
        
        #print "Endir(" + port + " , " + str( JTAG_STATES[regmatch.group( 1 )] ) + ");\n"
        Endir( port , JTAG_STATES[regmatch.group( 1 )] ) 

def EnddrHandler( _port, _line ):
        port = _port
        line = _line
        #portName = port.port

        print line
        
        regmatch = re.search( r"\s+(.*)\s*;", line, re.IGNORECASE )
        
        #print "Enddr(" + port + " , " + str( JTAG_STATES[regmatch.group( 1 )] ) + ");\n"
        Enddr( port , JTAG_STATES[regmatch.group( 1 )] )

def StateHandler( _port, _line ):
        ''' '''
        port = _port
        line = _line
        #portName = port.port

        print line
        
        regmatch = re.search( r"\s+(.*)\s*;", line, re.IGNORECASE )
        
        #print "GoToState(" + port + " , " + str( JTAG_STATES[regmatch.group( 1 ).upper()] ) + ");\n"
        GoToState( port , JTAG_STATES[regmatch.group( 1 ).upper()] )

def substr( string, offset, count ):
        if offset >= 0 and count >= 0:
                return string[offset : offset + count]
        
        elif offset < 0 and count >= 0:
                offset = len( string ) + offset
                return string[offset : offset + count ]
        
        elif offset < 0 and count < 0:
                offset = len( string ) + offset
                return string[offset : len( string ) + count]
        
        elif offset >= 0 and count < 0:
                return string[offset : len( string ) + count]


#####################################

serport = OpenPort( portName )

EchoOff( serport );
RunSVF( serport, FileName ) ;

serport.close();
